src/detect-parse.c

```diff
@@ -1735,6 +1735,84 @@ int DetectSignatureAddTransform(Signature *s, int transform, void *options)
     SCReturnInt(0);
 }
 
+// alprotos parameter is expected as an array terminated by ALPROTO_UNKNOWN
+int DetectSignatureSetMultiAppProto(Signature *s, const AppProto *alprotos)
+{
+    if (s->alproto != ALPROTO_UNKNOWN) {
+        // One alproto was set, check if it matches the new ones proposed
+        while (*alprotos != ALPROTO_UNKNOWN) {
+            if (s->alproto == *alprotos) {
+                // alproto already set to only one
+                return 0;
+            }
+            alprotos++;
+        }
+        // alproto already set and not matching the new set of alprotos
+        return -1;
+    }
+    if (s->init_data->alprotos[0] != ALPROTO_UNKNOWN) {
+        // check intersection of already used alprotos and new ones
+        for (size_t i = 0; i < SIG_ALPROTO_MAX; i++) {
+            if (s->init_data->alprotos[i] == ALPROTO_UNKNOWN) {
+                break;
+            }
+            // first disable the ones that do not match
+            bool found = false;
+            while (*alprotos != ALPROTO_UNKNOWN) {
+                if (s->init_data->alprotos[i] == *alprotos) {
+                    found = true;
+                    break;
+                }
+                alprotos++;
```

Comment:
This looks like we'll exhaust alprotos on checks against the first alproto entry of the signature if there was no match. Is this the intention?

Reply:
Not sure I follow what you mean here...

There is a double loop with the current signature alprotos, and the new ones...

Reply:
I mean for a say, `s->init_data->alprotos = [proto1, proto2, proto3, ALPROTO_UNKNOWN]`
and `alprotos = [proto4, proto3, ALPROTO_UNKNOWN]` coming to this part of the code,

First pass of the for loop looking for `proto1` will bring `alprotos` pointer out of the array. We won't go over `proto2` and `proto3` in the `s->init_data->alprotos`, is this possible/intended?

Reply:
Very right

Reply:
Thanks for spotting this